types/ziggy-js/index.d.ts

```diff
@@ -0,0 +1,56 @@
+// Type definitions for ziggy-js 0.9
+// Project: https://github.com/tightenco/ziggy#readme
+// Definitions by: Ben Allfree <https://github.com/benallfree>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.6
+
+export type EncodeURIParamType = string | number | boolean;
+
+export type NormalizedParams =
+    | {
+          [key: string]: EncodeURIParamType;
+      }
+    | EncodeURIParamType[];
+
+export type InputParams = NormalizedParams | EncodeURIParamType;
+
```

Comment:
@bakerkretzmar Taking a slightly different approach here, I noticed that we can be a little more specific about param data that has already been normalized. In all cases, it seems that param values must be something that `encodeURIComponent` can accept. In the Typescript world, that is the `EncodeURIParamType`. Notice that `object` types are not allowed (and in fact aren't supported by encodeURIComponent). But also, `array` types are not allowed even though they are technically supported by the encodeURIComponent implementation (at least in Chrome). According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) really only `string` should be allowed, but other types seem to work. I trust the official Typescript definition.

Reply:
Ziggy processes the params internally and will only pass integers and strings to `encodeURIComponent()`, so it is actually valid to pass an array of mixed integers, strings, and objects, or an object containing string keys and object values.

I know that's extremely lenient and hard to type-check, but that's how Ziggy works 😂 

All of the following are allowed as `params` to the `route()` function:

```js
1
'foo'
[1]
[1, 2]
['foo', 3]
{ city: 'Toronto', user: 'Jacob' }
{ city: 'Toronto', user: { id: 12, name: 'Jacob' } }
{ id: 1, name: 'Jacob' }
[{ id: 1, name: 'Jacob' }]
[{ id: 1, name: 'Jacob' }, 'foo']
[{ id: 1, name: 'Jacob' }, { id: 2, name: 'Ben' }]
[{ id: 3, name: 'Jacob', details: [/* literally anything is allowed here, it all gets thrown away */] }]
```

Reply:
I think we have this dialed in this time 👍   It will require the `id` key if it is a param value and will allow any key name if it is a param object. Plus it will allow strings/numbers/arrays thereof.